src/System.Management.Automation/namespaces/FileSystemProvider.cs

```diff
@@ -88,16 +88,25 @@ public FileSystemProvider()
         };
 
         /// <summary>
-        /// Converts all / in the path to \
+        /// Converts all / in the path to \ and on Windows shortpaths to longpaths
         /// </summary>
         /// <param name="path">
         /// The path to normalize.
         /// </param>
         /// <returns>
-        /// The path with all / normalized to \
+        /// The path with all / normalized to \ and on Windows a longpath
         /// </returns>
         private static string NormalizePath(string path)
         {
+            if (string.IsNullOrEmpty(path))
+            {
+                return path;
+            }
+
+            // Use FileInfo to convert short paths to long paths on Windows
+#if !UNIX
+            path = (new FileInfo(path)).FullName;
+#endif
```

Comment:
Do we really need this? Why we want short names?
I'd follow _new .Net Core behavior_ -  if they can do the "breaking change" we can do too.
The change can slow down PowerShell.

I think a fix should be in Environment.GetEnvironmentVariable
Also I guess that FileInfo return normalized path https://source.dot.net/#System.IO.FileSystem/System/IO/FileInfo.cs,35

Reply:
That line is so that we _don't_ have short paths.  Given a short path, it will resolve to the long path.

Reply:
`Path.GetFullPath` will avoid allocating a `FileInfo` (which just calls that anyway). The `GetFullPath` short path behavior isn't strictly correct btw. it is hugely expensive to check for the long path versions so we don't bother unless we see a tilde in the path. You can, in theory, have a short name that doesn't have a tilde- although I've never seen a real-world occurrence.

I'd recommend that you yank this method completely and straight up call `Path.GetFullPath`. You're going to get inconsistent results otherwise as we collapse separators in `GetFullPath` (which won't happen on Unix here). In addition the replacement here won't do the right thing if UNCs or device paths (e.g. `\\?\`) come in.

Reply:
> That line is so that we don't have short paths. Given a short path, it will resolve to the long path.

Currently PowerShell allow to process short paths and _preserve_ them. Question is - have/should we keep the behavior or not? I see we don't (in the case we should fix Intellisense too). I think it is for PowerShell Committee.
Also if we want to accept new .Net Core 3.0 behavior I don't see reasons to make exclusion for Environment.GetEnvironmentVariable() - what is backward compatibility for PowerShell scripts we warry about?

Reply:
@JeremyKuhne thanks for the tip on `Path.GetFullPath`, will make that change.  As for the slashes, this is existing code.  UNC detection happens somewhere else.

Reply:
> UNC detection happens somewhere else.

Curious as to where that is. It would be nice if device syntax was properly handled with UNCs. `\\?\UNC\server\share\` is a UNC while `\\?\C:\foo` isn't. :) I've got the logic for handling those cases in .NET.

Reply:
@JeremyKuhne it's a pretty [simple check](https://github.com/PowerShell/PowerShell/blob/d80154430d600a52a1d1e3a9f3ecf032b36185d4/src/System.Management.Automation/namespaces/FileSystemProvider.cs#L4715)

Reply:
Yeah, that would pull in things that aren't UNCs.  The logic we use is here: https://github.com/dotnet/corefx/blob/master/src/Common/src/CoreLib/System/IO/PathInternal.Windows.cs#L143-L188

The correct way to check would be something like `(path.StartsWith("\\\\", StringComparison.Ordinal) && !IsDevice(path)) || IsDeviceUnc(path)`. Or even more correct:

``` C#
path.Length > 2
    && ((IsDirectorySeparator(path[0]) && IsDirectorySeparator(path[1]) && !IsDevice(path)) || IsDeviceUnc(path))
```

Win32 doesn't care about the slash direction- they'll get normalized to backslash correctly. You can see this in action with `dir "//localhost/c$"`.

The values in handling device paths correctly:

1. Allows access to files you can't access otherwise
2. Allows long path access without manifesting & setting reg keys
3. Skips normalization (in the `\\?\` case), which will improve performance a bit

Reply:
@JeremyKuhne thanks for the feedback.  Looking again in the code, it appears we have a different [utility function](https://github.com/PowerShell/PowerShell/blob/d80154430d600a52a1d1e3a9f3ecf032b36185d4/src/System.Management.Automation/engine/Utils.cs#L966) relying on the uri class to determine if unc.  We should update usage of that other suspect api to use this one.  I'll take care of that as a separate PR from this one.

Reply:
The UNC check is fixed via https://github.com/PowerShell/PowerShell/pull/8680

Reply:
Perhaps I skipped something but I still think that we could:
```
#if !UNIX 
            return Path.GetFullPath(path); 
#endif
```